[BUG] Server can't be run due to lack of  Send trait on run() future

**Description**
The following simple code snippet fails on compilation when trying to use saphir 2.8.0 alongside with tokio 0.3.4
```rust
use saphir::server::Server;
use tokio::runtime::Runtime;

fn main() {
    let server = Server::builder().build();
    tokio::spawn(server.run());
}
```
```toml
[package]
name = "saphir-test"
version = "0.1.0"
edition = "2018"

[dependencies]
tokio = { version = "=0.3.4", features = [ "rt" ]}
saphir = "=2.8"
```

This code fails due to the following error:
```
error[E0277]: `(dyn Future<Output = ()> + Unpin + 'static)` cannot be sent between threads safely
   --> src\main.rs:6:5
    |
6   |     tokio::spawn(server.run());
    |     ^^^^^^^^^^^^ `(dyn Future<Output = ()> + Unpin + 'static)` cannot be sent between threads safely
    | 
   ::: C:\Users\pacmancoder\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-0.3.4\src\task\spawn.rs:128:21
    |
128 |         T: Future + Send + 'static,
    |                     ---- required by this bound in `tokio::spawn`
    |
    = help: the trait `Send` is not implemented for `(dyn Future<Output = ()> + Unpin + 'static)`
    = note: required because of the requirements on the impl of `Send` for `Unique<(dyn Future<Output = ()> + Unpin + 'static)>`
    = note: required because it appears within the type `Box<(dyn Future<Output = ()> + Unpin + 'static)>`
    = note: required because it appears within the type `Pin<Box<(dyn Future<Output = ()> + Unpin + 'static)>>`
    = note: required because it appears within the type `saphir::server::ServerShutdown`
    = note: required because it appears within the type `for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10, 't11, 't12, 't13, 't14, 't15, 't16, 't17, 't18, 't19, 't20, 't21, 't22, 't23, 't24, 't25, 't26, 't27, 't28, 't29, 't30, 't31, 't32, 't33, 't34, 't35, 't36, 't37, 't38, 't39, 't40, 't41, 't42, 't43, 't44, 't45, 't46, 't47, 't48, 't49, 't50, 't51, 't52, 't53, 't54, 't55, 't56, 't57, 't58> {ResumeTy, saphir::server::Server, ListenerConfig, saphir::server::Stack, HeaderValue, Option<usize>, &'r saphir::server::Stack, saphir::hyper::server::conn::Http, &'s String, String, impl Future, (), tokio::net::tcp::listener::TcpListener, SocketAddr, tokio::net::tcp::incoming::Incoming<'t0>, saphir::server::ServerShutdown, Arc<saphir::server::SeverShutdownState>, Option<u64>, u64, futures_util::stream::stream::for_each_concurrent::ForEachConcurrent<tokio::net::tcp::incoming::Incoming<'t1>, impl Future, [closure@saphir::server::Server::run::{closure#0}::{closure#1}]>, saphir::server::ServerFuture<futures_util::stream::stream::for_each_concurrent::ForEachConcurrent<tokio::net::tcp::incoming::Incoming<'t17>, impl Future, [closure@saphir::server::Server::run::{closure#0}::{closure#1}]>, saphir::server::ServerShutdown>, futures_util::stream::stream::for_each_concurrent::ForEachConcurrent<tokio::net::tcp::incoming::Incoming<'t33>, impl Future, [closure@saphir::server::Server::run::{closure#0}::{closure#2}]>, saphir::server::ServerFuture<futures_util::stream::stream::for_each_concurrent::ForEachConcurrent<tokio::net::tcp::incoming::Incoming<'t46>, impl Future, [closure@saphir::server::Server::run::{closure#0}::{closure#2}]>, saphir::server::ServerShutdown>}`
    = note: required because it appears within the type `[static generator@saphir::server::Server::run::{closure#0} for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10, 't11, 't12, 't13, 't14, 't15, 't16, 't17, 't18, 't19, 't20, 't21, 't22, 't23, 't24, 't25, 't26, 't27, 't28, 't29, 't30, 't31, 't32, 't33, 't34, 't35, 't36, 't37, 't38, 't39, 't40, 't41, 't42, 't43, 't44, 't45, 't46, 't47, 't48, 't49, 't50, 't51, 't52, 't53, 't54, 't55, 't56, 't57, 't58> {ResumeTy, saphir::server::Server, ListenerConfig, saphir::server::Stack, HeaderValue, Option<usize>, &'r saphir::server::Stack, saphir::hyper::server::conn::Http, &'s String, String, impl Future, (), tokio::net::tcp::listener::TcpListener, SocketAddr, tokio::net::tcp::incoming::Incoming<'t0>, saphir::server::ServerShutdown, Arc<saphir::server::SeverShutdownState>, Option<u64>, u64, futures_util::stream::stream::for_each_concurrent::ForEachConcurrent<tokio::net::tcp::incoming::Incoming<'t1>, impl Future, [closure@saphir::server::Server::run::{closure#0}::{closure#1}]>, saphir::server::ServerFuture<futures_util::stream::stream::for_each_concurrent::ForEachConcurrent<tokio::net::tcp::incoming::Incoming<'t17>, impl Future, [closure@saphir::server::Server::run::{closure#0}::{closure#1}]>, saphir::server::ServerShutdown>, futures_util::stream::stream::for_each_concurrent::ForEachConcurrent<tokio::net::tcp::incoming::Incoming<'t33>, impl Future, [closure@saphir::server::Server::run::{closure#0}::{closure#2}]>, saphir::server::ServerFuture<futures_util::stream::stream::for_each_concurrent::ForEachConcurrent<tokio::net::tcp::incoming::Incoming<'t46>, impl Future, [closure@saphir::server::Server::run::{closure#0}::{closure#2}]>, saphir::server::ServerShutdown>}]`
    = note: required because it appears within the type `from_generator::GenFuture<[static generator@saphir::server::Server::run::{closure#0} for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10, 't11, 't12, 't13, 't14, 't15, 't16, 't17, 't18, 't19, 't20, 't21, 't22, 't23, 't24, 't25, 't26, 't27, 't28, 't29, 't30, 't31, 't32, 't33, 't34, 't35, 't36, 't37, 't38, 't39, 't40, 't41, 't42, 't43, 't44, 't45, 't46, 't47, 't48, 't49, 't50, 't51, 't52, 't53, 't54, 't55, 't56, 't57, 't58> {ResumeTy, saphir::server::Server, ListenerConfig, saphir::server::Stack, HeaderValue, Option<usize>, &'r saphir::server::Stack, saphir::hyper::server::conn::Http, &'s String, String, impl Future, (), tokio::net::tcp::listener::TcpListener, SocketAddr, tokio::net::tcp::incoming::Incoming<'t0>, saphir::server::ServerShutdown, Arc<saphir::server::SeverShutdownState>, Option<u64>, u64, futures_util::stream::stream::for_each_concurrent::ForEachConcurrent<tokio::net::tcp::incoming::Incoming<'t1>, impl Future, [closure@saphir::server::Server::run::{closure#0}::{closure#1}]>, saphir::server::ServerFuture<futures_util::stream::stream::for_each_concurrent::ForEachConcurrent<tokio::net::tcp::incoming::Incoming<'t17>, impl Future, [closure@saphir::server::Server::run::{closure#0}::{closure#1}]>, saphir::server::ServerShutdown>, futures_util::stream::stream::for_each_concurrent::ForEachConcurrent<tokio::net::tcp::incoming::Incoming<'t33>, impl Future, [closure@saphir::server::Server::run::{closure#0}::{closure#2}]>, saphir::server::ServerFuture<futures_util::stream::stream::for_each_concurrent::ForEachConcurrent<tokio::net::tcp::incoming::Incoming<'t46>, impl Future, [closure@saphir::server::Server::run::{closure#0}::{closure#2}]>, saphir::server::ServerShutdown>}]>`
    = note: required because it appears within the type `impl Future`
    = note: required because it appears within the type `impl Future`

error: aborting due to previous error; 1 warning emitted
```

However, on version 2.7.7 of saphir everything works as expected.

**Expected behavior**
The provided snippet can be run with saphir 2.8.0 the same way it works with saphir 2.7.7

**Additional info (please complete the following information):**
 - OS & OS Version: Windows 10 rev.2004
 - Saphir Version: 2.8.0
 - Rust toolchain & version: stable 1.48

@Samuel-B-D @richerarc tried to run the same snippet which I posted above, same result with tokio 0.2.22:
```toml
[package]
name = "saphir-test"
version = "0.1.0"
edition = "2018"

[dependencies]
tokio = { version = "=0.2.22", futures = ["rt-core"]}
saphir = "2.8.0"
```

Error:
```
error[E0277]: `(dyn std::future::Future<Output = ()> + Unpin + 'static)` cannot be sent between threads safely
   --> src\main.rs:6:5
    |
6   |     tokio::spawn(server.run());
    |     ^^^^^^^^^^^^ `(dyn std::future::Future<Output = ()> + Unpin + 'static)` cannot be sent between threads safely
    | 
   ::: C:\Users\pacmancoder\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-0.2.22\src\task\spawn.rs:127:21
    |
127 |         T: Future + Send + 'static,
    |                     ---- required by this bound in `tokio::spawn`
```

--- a/saphir/src/middleware.rs
+++ b/saphir/src/middleware.rs
@@ -171,7 +171,7 @@ impl MiddlewareChain for MiddleChainEnd {
     #[inline]
     fn next(&self, mut ctx: HttpContext) -> BoxFuture<'static, Result<HttpContext, SaphirError>> {
         async {
-            let router = ctx.router.take().ok_or_else(|| SaphirError::Internal(InternalError::Stack))?;
+            let router = ctx.router.take().ok_or(SaphirError::Internal(InternalError::Stack))?;
             router.dispatch(ctx).await
         }
         .boxed()

--- a/saphir/src/multipart/parser.rs
+++ b/saphir/src/multipart/parser.rs
@@ -132,8 +132,7 @@ fn headers(input: &[u8]) -> IResult<&[u8], FieldHeaders> {
         }
     }
 
-    let (content_disposition_name, content_disposition_filename) =
-        content_disposition.ok_or_else(|| nom::Err::Error((input, nom::error::ErrorKind::MapOpt)))?;
+    let (content_disposition_name, content_disposition_filename) = content_disposition.ok_or(nom::Err::Error((input, nom::error::ErrorKind::MapOpt)))?;
 
     Ok((
         input,

--- a/saphir/src/server.rs
+++ b/saphir/src/server.rs
@@ -217,7 +217,7 @@ impl ListenerBuilder {
             request_timeout_ms,
             server_name: server_name.unwrap_or_else(|| DEFAULT_SERVER_NAME.to_string()),
             request_body_max,
-            shutdown
+            shutdown,
         }
     }
 }

--- a/saphir/src/server.rs
+++ b/saphir/src/server.rs
@@ -76,7 +76,7 @@ pub struct ListenerBuilder {
     cert_config: Option<SslConfig>,
     #[cfg(feature = "https")]
     key_config: Option<SslConfig>,
-    shutdown_signal: Option<Box<dyn Future<Output = ()> + Unpin + 'static>>,
+    shutdown_signal: Option<Box<dyn Future<Output = ()> + Unpin + Send + 'static>>,
     graceful_shutdown: bool,
 }
 
@@ -131,7 +131,7 @@ impl ListenerBuilder {
     /// request to be completed before shutting down but will stop accepting
     /// new requests.
     #[inline]
-    pub fn shutdown<F: Future<Output = ()> + Unpin + 'static>(mut self, signal: F, graceful: bool) -> Self {
+    pub fn shutdown<F: Future<Output = ()> + Unpin + Send + 'static>(mut self, signal: F, graceful: bool) -> Self {
         self.shutdown_signal = Some(Box::new(signal));
         self.graceful_shutdown = graceful;
         self
@@ -330,11 +330,11 @@ impl SeverShutdownState {
 struct ServerShutdown {
     graceful: bool,
     state: Arc<SeverShutdownState>,
-    signal: Pin<Box<dyn Future<Output = ()> + Unpin + 'static>>,
+    signal: Pin<Box<dyn Future<Output = ()> + Unpin + Send + 'static>>,
 }
 
 impl ServerShutdown {
-    pub fn new<F: Future<Output = ()> + Unpin + 'static>(graceful: bool, signal: F) -> Self {
+    pub fn new<F: Future<Output = ()> + Unpin + Send + 'static>(graceful: bool, signal: F) -> Self {
         ServerShutdown {
             graceful,
             state: Arc::new(Default::default()),

